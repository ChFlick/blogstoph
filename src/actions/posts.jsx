import database from '../firebase/firebase';

export const addPost = (post) => ({
    type: 'ADD_POST',
    post
});

export const startAddPost = (post) => {
    const { title, content, author, date, published } = post;

    return (dispatch, getState) => {
        const visibility = published ? 'public' : 'private';
        return database.ref('posts/' + visibility).push({ title, content, author, date }).then((ref) => {
            dispatch(addPost({
                id: ref.key,
                published,
                ...post
            }));
        });
    };
};

export const editPost = (id, updates) => ({
    type: 'EDIT_POST',
    id,
    updates
});

export const startEditPost = (id, post) => {
    const { title, content, author, date, published } = post;

    return (dispatch, getState) => {
        const prevPostPublished = getState().posts.find((post) => post.id === id).published;
        const visibility = published ? 'public' : 'private';

        if (prevPostPublished !== published) {
            const prevPostVisibility = prevPostPublished ? 'public' : 'private';
            database.ref(`posts/${prevPostVisibility}/${id}`).remove();
        }

        return database.ref(`posts/${visibility}/${id}`).set({ title, content, author, date }).then((ref) => {
            dispatch(editPost(
                id,
                post
            ));
        });
    };
};

export const setPosts = (posts) => ({
    type: 'SET_POSTS',
    posts
});

export const startSetPosts = () => {
    const addPostFromSnapshot = (published, posts) => (postSnapshot) => {
        posts.push({
            id: postSnapshot.key,
            published: !!published,
            ...postSnapshot.val()
        });
    };

    return (dispatch, getState) => {
        return Promise.all(getPostsValueDBReferenceByAuth(getState)
        ).then(([publicPostsSnapshot, privatePostsSnapshot]) => {
            const posts = [];
            publicPostsSnapshot.forEach(addPostFromSnapshot(true, posts));

            if (privatePostsSnapshot) {
                privatePostsSnapshot.forEach(addPostFromSnapshot(false, posts));
            }

            dispatch(setPosts(posts));
        });
    };
};

function getPostsValueDBReferenceByAuth(getState) {
    return (getState().auth && getState().auth.uid) ?
        [database.ref('posts/public').once('value'), database.ref('posts/private').once('value')] :
        [database.ref('posts/public').once('value')];
}

