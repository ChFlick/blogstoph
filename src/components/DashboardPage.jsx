import React, { Fragment } from 'react';

import PostList from './PostList';

const DashboardPage = () => (
    <Fragment>
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title page-header__title--xl">Blogstoph</h1>
                <hr className="page-header__bar"/>
                <h2 className="page-header__subtitle">Christoph Flicks IT blog</h2>
            </div>
        </div>
        <PostList />
    </Fragment>
);

export default DashboardPage;
