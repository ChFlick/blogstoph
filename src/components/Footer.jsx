import React, { Fragment } from 'react';

const Footer = (props) => (
    <Fragment>
        <hr className="bar--light-grey" />
        <div className="copyright">Copyright Â© Christoph Flick 2018</div>
    </Fragment>
);

export default Footer;
