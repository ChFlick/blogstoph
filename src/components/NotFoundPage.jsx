import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
    <Fragment>
        <header className="page-header">
            <h1>404 Not Found</h1>
        </header>
        <section className="content-container align-center">
            <p>The page you were looking for could not be found.</p>
            <Link to="/">Click here to go back to the home page</Link>
        </section>
    </Fragment>
);

export default NotFoundPage;
