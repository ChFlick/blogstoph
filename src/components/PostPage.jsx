import React, { Fragment } from 'react';
import Remarkable from 'react-remarkable';
import { connect } from 'react-redux';

export const PostPage = ({post}) => (
    <Fragment>
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title page-header__title--xl">{post.title}</h1>
                <h2 className="page-header__subtitle"></h2>
            </div>
        </div>
        <Remarkable>
            {post.content}
        </Remarkable>
    </Fragment>
);

const mapStateToProps = (state, props) => ({
    post: state.posts.find((post) => post.id === props.match.params.id)
});

export default connect(mapStateToProps, undefined)(PostPage);
