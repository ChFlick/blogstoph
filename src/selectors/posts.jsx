const getVisiblePosts = (posts = [], { published, sortBy = 'date' } = {}) => {
    return posts.filter((post) => {
        const publishedMatch = published === undefined || (!!published === post.published);

        return publishedMatch;
    }).sort((a, b) => {
        return b[sortBy] - a[sortBy];
    });
};

const getPostByIdOrTitle = (posts, searchField) => {
    return posts.find((post) =>
        post.id === searchField ||
        post.title.toLowerCase().replace(/ /g, '-') === searchField.toLowerCase());
};

export { getVisiblePosts, getPostByIdOrTitle };
