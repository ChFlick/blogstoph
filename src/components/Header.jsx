import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from '../actions/auth';

const DEFAULT_HEADER_CLASSES = 'header header--fixed-top';

export class Header extends React.Component {
    state = {
        headerClasses: DEFAULT_HEADER_CLASSES
    }

    componentDidMount = () => {
        window.addEventListener('scroll', this.handleScroll);
    };

    componentWillUnmount = () => {
        window.removeEventListener('scroll', this.handleScroll);
    };

    handleScroll = () => {
        const scrollPosPrev = this.state.scrollPos;
        const scrollPos = window.scrollY;
        let headerClasses = 'header header--fixed-top';

        if (scrollPos > 80) {
            headerClasses = 'header header--fixed';
            if (scrollPos < scrollPosPrev) {
                headerClasses += ' header--visible';
            }

            this.setState(() => ({ headerClasses }));
        } else if (scrollPos === 0) {
            this.setState(() => ({ headerClasses: DEFAULT_HEADER_CLASSES }));
        }

        this.setState(() => ({ scrollPos }));
    };

    createAreaSwitchButton = () =>
        this.props.private ?
            <Link className="button button--link button--header" to="/dashboard">Visitor Area</Link> :
            <Link className="button button--link button--header" to="/editor/dashboard">Editor Area</Link>;


    render() {
        return (
            <header id="header" className={this.state.headerClasses} onScroll={this.handleScroll}>
                <div className="content-container">
                    <div className="header__content">
                        <Link className="header__title" to="/dashboard">
                            <h1>Blogstoph</h1>
                        </Link>
                        {this.props.isAuthenticated ? (
                            <div>
                                {this.createAreaSwitchButton()}
                                <button className="button button--link button--header" onClick={this.props.startLogout}>Logout</button>
                            </div>
                        ) : (
                                <Link className="button button--link" to="/login">Login</Link>
                            )}
                    </div>
                </div>
            </header>
        );
    };
};

const mapStateToProps = (state) => ({
    isAuthenticated: !!state.auth.uid
});

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
