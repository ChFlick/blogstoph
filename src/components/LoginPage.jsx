import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Blogstoph Login</h1>
            <p>Log in to edit your posts.</p>
            <button className="button" onClick={startLogin}>Login with Google</button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin('google'))
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
