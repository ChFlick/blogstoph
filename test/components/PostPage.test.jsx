import React from 'react';
import { shallow } from 'enzyme';
import { publishedPosts } from '../fixtures/posts';

import { PostPage } from '../../src/components/PostPage';

test('should render one post page correctly', () => {
    const wrapper = shallow(<PostPage post={publishedPosts[0]}/>);
    expect(wrapper).toMatchSnapshot();
});
