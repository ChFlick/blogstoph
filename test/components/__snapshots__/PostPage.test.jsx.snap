// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render one post page correctly 1`] = `
<React.Fragment>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title page-header__title--xl"
      >
        Some Title
      </h1>
      <h2
        className="page-header__subtitle"
      />
    </div>
  </div>
  <Remarkable
    container="div"
    options={Object {}}
  >
    This is a Content of a Post...
  </Remarkable>
</React.Fragment>
`;
